src/ray/raylet/node_manager.cc

```diff
@@ -2152,6 +2152,8 @@ void NodeManager::AsyncGetOrWait(const std::shared_ptr<ClientConnection> &client
   std::shared_ptr<WorkerInterface> worker = worker_pool_.GetRegisteredWorker(client);
   if (!worker) {
     worker = worker_pool_.GetRegisteredDriver(client);
+  } else if (worker->GetGrantedLeaseId().IsNil()) {
```

Comment:
Checking lease id being Nil is not complete for checking that the worker is no longer processing the task because:

1. The worker might be leased out again with a new lease ID at this point
2. The worker is reused by the caller to execute another task

On raylet side, we can know 1 but we won't know 2 so there will be cases we register an unnecessary get request but it will be cancelled later.

Another fix we can do is that we always register here but remove the early return check (`worker->GetGrantedLeaseId().IsNil()`) in `HandleDirectCallTaskUnblocked` and rely on the fact that core worker guarantees to generate these pair of rpcs.